I get the following error when running the sessions test. Please help.
```bash
2022-06-05 21:35:28,680 240 ERROR db_test odoo.addons.openacademy.tests.test_openacademy_session: ERROR: TestSession.test_10_create_session
Traceback (most recent call last):
  File "/mnt/extra-addons/openacademy/tests/test_openacademy_session.py", line 33, in test_10_create_session
    session = self.create_session(**session_data)
  File "/mnt/extra-addons/openacademy/tests/commontest.py", line 37, in create_session
    session = self.Session.create({
  File "<decorator-gen-67>", line 2, in create
  File "/usr/lib/python3/dist-packages/odoo/api.py", line 412, in _model_create_multi
    return create(self, [arg])
  File "/usr/lib/python3/dist-packages/odoo/addons/base/models/ir_fields.py", line 613, in create
    recs = super().create(vals_list)
  File "<decorator-gen-13>", line 2, in create
  File "/usr/lib/python3/dist-packages/odoo/api.py", line 413, in _model_create_multi
    return create(self, arg)
  File "/usr/lib/python3/dist-packages/odoo/models.py", line 4070, in create
    records = self._create(data_list)
  File "/usr/lib/python3/dist-packages/odoo/models.py", line 4236, in _create
    records.check_access_rule('create')
  File "/usr/lib/python3/dist-packages/odoo/models.py", line 3553, in check_access_rule
    invalid = self - self._filter_access_rules_python(operation)
  File "/usr/lib/python3/dist-packages/odoo/models.py", line 3608, in _filter_access_rules_python
    return self.sudo().filtered_domain(dom or [])
  File "/usr/lib/python3/dist-packages/odoo/models.py", line 5516, in filtered_domain
    field = model._fields[fname]
KeyError: 'responsible_id'
```

--- a/openacademy/security/openacademy_security.xml
+++ b/openacademy/security/openacademy_security.xml
@@ -15,13 +15,13 @@
 
     <record id="restricts_responsible" model="ir.rule">
         <field name="name">Restricts write and unlink accesses.</field>
-        <field name="model_id" ref="model_openacademy_session"/>
+        <field name="model_id" ref="model_openacademy_course"/>
         <field name="groups" eval="[(4, ref('group_openacademy_manager'))]"/>
-        <field name="perm_read" eval="True"/>
-        <field name="perm_write" eval="False"/>
-        <field name="perm_create" eval="True"/>
-        <field name="perm_unlink" eval="False"/>
-        <field name="domain_force">[('responsible_id','=', False)]</field>
+        <field name="perm_read" eval="False"/>
+        <field name="perm_write" eval="True"/>
+        <field name="perm_create" eval="False"/>
+        <field name="perm_unlink" eval="True"/>
+        <field name="domain_force">[('responsible_id','=', uid)]</field>
     </record>
 
 </odoo>

--- a/openacademy/tests/__init__.py
+++ b/openacademy/tests/__init__.py
@@ -1,0 +1,2 @@
+from . import test_openacademy_course
+from . import test_openacademy_session

--- a/openacademy/tests/commontest.py
+++ b/openacademy/tests/commontest.py
@@ -1,0 +1,44 @@
+from odoo.tests import common
+
+
+class CommonTest(common.TransactionCase):
+    """Test Base class from which all tests inherit."""
+
+    def setUp(self, *args, **kwargs):
+        super().setUp(*args, **kwargs)
+        self.user_admin = self.env.ref("base.user_admin")
+        self.Course = self.env["openacademy.course"].with_user(self.user_admin)
+        self.Session = self.env["openacademy.session"].with_user(
+            self.user_admin
+        )
+        self.Instructor = self.env["res.partner"].with_user(self.user_admin)
+
+    def create_course(self, name, description, responsible_id):
+        """Create and return a course object."""
+        course = self.Course.create({
+            "name": name,
+            "description": description,
+            "responsible_id": responsible_id,
+        })
+        return course
+
+    def create_instructor(self, name, instructor: bool):
+        """Create and return an instructor object."""
+        instructor = self.Instructor.create({
+            "name": name,
+            "instructor": instructor,
+        })
+        return instructor
+
+    def create_session(
+        self, name, number_seats, instructor_id, taken_seats, course_id
+    ):
+        """Create and return a session object."""
+        session = self.Session.create({
+            "name": name,
+            "number_seats": number_seats,
+            "instructor_id": instructor_id,
+            "taken_seats": taken_seats,
+            "course_id": course_id,
+        })
+        return session

--- a/openacademy/tests/test_openacademy_course.py
+++ b/openacademy/tests/test_openacademy_course.py
@@ -1,0 +1,77 @@
+from . commontest import CommonTest
+
+from psycopg2.errors import CheckViolation, UniqueViolation
+from odoo.tools import mute_logger
+
+
+class TestCourse(CommonTest):
+
+    def test_10_create_course(self):
+        """
+        Test create a course.
+        """
+        course_data = {
+            "name": "Test course",
+            "description": "This is a test description.",
+            "responsible_id": self.user_admin.id,
+        }
+        course = self.create_course(**course_data)
+
+        self.assertTrue(course.id, "This must be a valid id.")
+        self.assertEqual(
+            course.name, course_data["name"],
+            f"This must be equal to {course_data['name']}.",
+        )
+        self.assertEqual(
+            course.description, course_data["description"],
+            f"This must be equal to {course_data['description']}.",
+        )
+        self.assertEqual(
+            course.responsible_id.id, course_data["responsible_id"],
+            f"This must be equal to {course_data['responsible_id']}.",
+        )
+
+    @mute_logger("odoo.sql_db")
+    def test_20_description_and_title_different(self):
+        """
+        Test description and title different.
+        """
+        course_data = {
+            "name": "Test course",
+            "description": "Test course",
+            "responsible_id": self.user_admin.id,
+        }
+        with self.assertRaises(CheckViolation):
+            self.create_course(**course_data)
+
+    @mute_logger("odoo.sql_db")
+    def test_30_unique_course_name(self):
+        """
+        Test unique course name.
+        """
+        course_data = {
+            "name": "Test course",
+            "description": "Test description",
+            "responsible_id": self.user_admin.id,
+        }
+        self.create_course(**course_data)
+
+        with self.assertRaises(UniqueViolation):
+            self.create_course(**course_data)
+
+    def test_40_copy_course(self):
+        """
+        Test copy course.
+        """
+        course_data = {
+            "name": "Test course",
+            "description": "Test description",
+            "responsible_id": self.user_admin.id,
+        }
+        self.create_course(**course_data)
+        course = self.Course.search([("name", "=", course_data["name"])])
+        copy_course = course.copy()
+        self.assertEqual(
+            copy_course.name, f"Copy of {course_data['name']}.",
+            f"This must be equal to Copy of {course_data['name']}.",
+        )

--- a/openacademy/tests/test_openacademy_session.py
+++ b/openacademy/tests/test_openacademy_session.py
@@ -1,0 +1,117 @@
+from . commontest import CommonTest
+
+from odoo.exceptions import ValidationError
+from odoo.tools import mute_logger
+from odoo.tests import Form
+
+
+class TestSession(CommonTest):
+
+    def test_10_create_session(self):
+        """
+        Test create a session.
+        """
+        course_data = {
+            "name": "Test course",
+            "description": "This is a test description.",
+            "responsible_id": self.user_admin.id,
+        }
+        course = self.create_course(**course_data)
+
+        instructor_data = {
+            "name": "Test instructor",
+            "instructor": True,
+        }
+        instructor = self.create_instructor(**instructor_data)
+
+        session_data = {
+            "name": "Test session",
+            "number_seats": 10,
+            "instructor_id": instructor.id,
+            "taken_seats": 5,
+            "course_id": course.id,
+        }
+        session = self.create_session(**session_data)
+
+        self.assertTrue(session.id, "This must be a valid id.")
+        self.assertEqual(
+            session.name, session_data["name"],
+            f"This must be equal to {session_data['name']}.",
+        )
+        self.assertEqual(
+            session.number_seats, session_data["number_seats"],
+            f"This must be equal to {session_data['number_seats']}.",
+        )
+        self.assertEqual(
+            session.instructor_id.id, session_data["instructor_id"],
+            f"This must be equal to {session_data['instructor_id']}.",
+        )
+        self.assertEqual(
+            session.taken_seats, session_data["taken_seats"],
+            f"This must be equal to {session_data['taken_seats']}.",
+        )
+        self.assertEqual(
+            session.course_id.id, session_data["course_id"],
+            f"This must be equal to {session_data['course_id']}.",
+        )
+
+    @mute_logger("odoo.sql_db")
+    def test_20_check_instructor_cannot_be_attendee(self):
+        """
+        Test instructor cannot be attendee.
+        """
+        course_data = {
+            "name": "Test course",
+            "description": "This is a test description.",
+            "responsible_id": self.user_admin.id,
+        }
+        course = self.create_course(**course_data)
+
+        instructor_data = {
+            "name": "Test instructor",
+            "instructor": True,
+        }
+        instructor = self.create_instructor(**instructor_data)
+
+        session_data = {
+            "name": "Test session",
+            "instructor_id": instructor.id,
+            "course_id": course.id,
+            "attendee_ids": [(6, 0, [instructor.id])],
+        }
+        with self.assertRaises(ValidationError):
+            self.Session.create(session_data)
+
+    def test_30_compute_perc_taken_seats(self):
+        """
+        Test compute perc taken seats.
+        """
+        session_form = Form(self.Session)
+        session_form.number_seats = 10
+        session_form.taken_seats = 5
+
+        perc_taken_seats = (session_form.taken_seats / session_form.number_seats) * 100
+
+        self.assertEqual(
+            session_form.perc_taken_seats, perc_taken_seats,
+            f"This must be equal to {perc_taken_seats}.",
+        )
+
+    def test_40_number_seats_less_than_zero(self):
+        """
+        Test number_seats less than zero.
+        """
+        session_form = Form(self.Session)
+
+        with self.assertLogs(level="WARNING"):
+            session_form.number_seats = -5
+
+    def test_50_seats_taken_greater_than_number_of_seats(self):
+        """
+        Test taken_seats greater than number_seats.
+        """
+        session_form = Form(self.Session)
+
+        with self.assertLogs(level="WARNING"):
+            session_form.taken_seats = 20
+            session_form.number_seats = 10

